### PR TITLE
feat(core): add support for cjson_encode_number_precision

### PR DESCRIPTION
**Pull request by sabertobihwy**
This commit introduces a new configuration option called `cjson_encode_number_precision`. It specifies the precision used when encoding floating point numbers to JSON. The default value is 14, and the maximum value is 16. The commit also includes changes to various modules that make use of this configuration option.

### Checklist

- [x] The Pull Request has tests
- [x] There's an entry in the CHANGELOG
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix [FTI-5279](https://konghq.atlassian.net/browse/FTI-5279)

[FTI-5279]: https://konghq.atlassian.net/browse/FTI-5279?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by sabertobihwy** [comment]
There are some points to discuss:

- `kong/pdk/vault.lua` also has a `local cjson = require(“cjson.safe”).new()`, but it doesn’t use the` cjson.encode` function, so I didn’t handle it.
- If a module calls `require("cjson.safe").new()`, then the `cjson.encode_sparse_array(nil, nil, 2^15)` in `globalpatches.lua` will become ineffective. Is this the expected behavior?

File: kong/globalpatches.lua [hunk 1]
```diff
@@ -15,9 +15,11 @@ return function(options)
   local meta = require "kong.meta"
 
 
+  -- TODO: if other modules call require("cjson.safe").new()
```
Comment (bungle) [inline kong/globalpatches.lua:18]: I think we have that in other modules too.

Comment (sabertobihwy) [reply kong/globalpatches.lua:18]: yes, and I only found two in kong source code.

**Comment by sabertobihwy** [comment]
> It's not 100% air-tight, but it should cover most cases. We could even expose this in the PDK at _G.kong.json or something of that nature. Existing code that uses cjson will just need s/cjson/kong.tools.json/.

Reasonable.

I think it need to open another PR to handle this.

File: kong.conf.default [hunk 2]
```diff
@@ -1845,6 +1845,14 @@
                                   # valid UUID. When empty, node ID is
                                   # automatically generated.
 
+#cjson_encode_number_precision
+                                  # Precision used when encoding numbers to
+                                  # JSON. This is the maximum number of
+                                  # significant digits used when encoding
+                                  # floating point numbers to JSON text.
+                                  #
+                                  # The default value is 14, the maximum value is 16.
```
Comment (StarlightIbuki) [inline kong.conf.default:1854]:
```suggestion
                                  # The value is 14, 15, or 16, and defaults to 14.
```

File: kong/conf_loader/init.lua [hunk 3]
```diff
@@ -597,6 +597,8 @@ local CONF_PARSERS = {
   admin_gui_url = {typ = "string"},
   admin_gui_path = {typ = "string"},
   admin_gui_api_url = {typ = "string"},
+
+  cjson_encode_number_precision = { typ = "number", },
```
Comment (StarlightIbuki) [inline kong/conf_loader/init.lua:601]:
```suggestion
  cjson_encode_number_precision = { typ = "number", default = 14 },
```

File: kong/templates/kong_defaults.lua [hunk 4]
```diff
@@ -23,6 +23,7 @@ error_template_json = NONE
 error_template_xml = NONE
 error_template_plain = NONE
 node_id = NONE
+cjson_encode_number_precision = NONE
```
Comment (StarlightIbuki) [inline kong/templates/kong_defaults.lua:26]:
```suggestion
cjson_encode_number_precision = 14
```

File: spec/02-integration/05-proxy/33-cjson_encode_number_precision_spec.lua [hunk 5]
```diff
@@ -0,0 +1,154 @@
+local helpers = require "spec.helpers"
+
+for _, strategy in helpers.each_strategy() do
+    describe("cjson encode number with defalut(14) precision [#" .. strategy .. "]", function()
```
Comment (StarlightIbuki) [inline spec/02-integration/05-proxy/33-cjson_encode_number_precision_spec.lua:4]: Should we test all the precisions with a loop?

**Comment by ADD-SP** [comment]
Is there any context for this PR? Which issues is this PR going to fix? 

It pains me to add this configuration option, we don't want to add configuration options if we don't have to, because it would make configuring Kong more complicated.

**Review by StarlightIbuki — COMMENTED** [review]
I feel it is more direct and intuitive to specify the default value instead of relying on the libraries' default behavior.

**Review by ADD-SP — COMMENTED** [review]
Is there any context for this PR? Which issues is this PR going to fix?

It pains me to add this configuration option like this, we don't want to add configuration options if we don't have to, because it would make configuring Kong more complicated.

**Comment by ms2008** [comment]
> If a module calls require("cjson.safe").new(), then the cjson.encode_sparse_array(nil, nil, 2^15) in globalpatches.lua will become ineffective. Is this the expected behavior?

I think this issue should also be addressed in this PR to ensure that cjson behavior is consistent so as not to annoy users.

**Comment by sabertobihwy** [comment]
> Is there any context for this PR? Which issues is this PR going to fix?

background: https://konghq.atlassian.net/browse/FTI-5265?focusedCommentId=108202

Simply put, the default precision (14) cannot meet the needs of some users.

> It pains me to add this configuration option like this, we don't want to add configuration options if we don't have to, because it would make configuring Kong more complicated.

I also don’t like adding configurations. Another way is to increase the default precision from 14 to 16, but this is a breaking change, so I didn’t do it initially.

**Comment by sabertobihwy** [comment]
> I think this issue should also be addressed in this PR to ensure that cjson behavior is consistent so as not to annoy users.

I think it would be better to complete this with another PR that matches the solution proposed by @flrgh , otherwise this PR will lose its original purpose and become bloated.

File: kong/conf_loader/init.lua (hunk 3 above)
Comment (sabertobihwy) [reply kong/conf_loader/init.lua:601]: all items in this table did not set `default`, so I follow this.

File: spec/02-integration/05-proxy/33-cjson_encode_number_precision_spec.lua (hunk 5 above)
Comment (sabertobihwy) [reply spec/02-integration/05-proxy/33-cjson_encode_number_precision_spec.lua:4]: These two tests cover precision 14 and 16, but there is no test for 15. I think we can be more optimistic.

File: kong/templates/kong_defaults.lua (hunk 4 above)
Comment (sabertobihwy) [reply kong/templates/kong_defaults.lua]: done

File: kong.conf.default (hunk 2 above)
Comment (sabertobihwy) [reply kong.conf.default]: done

**Comment by sabertobihwy** [comment]
> I feel it is more direct and intuitive to specify the default value instead of relying on the libraries' default behavior.

Makes sense, should we declare a default precision of 14 in `globalpatchers`?

**Comment by StarlightIbuki** [comment]
> > I feel it is more direct and intuitive to specify the default value instead of relying on the libraries' default behavior.
> 
> Makes sense, should we declare a default precision of 14 in `globalpatchers`?

We can make it the default value of the configuration and always pass the argument.

**Comment by sabertobihwy** [comment]
> We can make it the default value of the configuration and always pass the argument.

That is a breaking change, and I am concerned because not all users require 16-bit precision.

**Comment by sabertobihwy** [comment]
So now there is a disagreement: whether to upgrade the default precision from 14 digits to 16 digits. 
My opinion is no. 
What is your opinion? cc @StarlightIbuki @ADD-SP @bungle

**Comment by StarlightIbuki** [comment]
> So now there is a disagreement: whether to upgrade the default precision from 14 digits to 16 digits. My opinion is no. What is your opinion? cc @StarlightIbuki @ADD-SP @bungle

I did not mean we need 16 as default. I mean 14 as the default but explicitly passed to the lib.

**Comment by vm-001** [comment]
Please add a changelog file

**Comment by kikito** [comment]
After considernig this in PR queue review, we are closing this. It is not a burning issue, it's controversial and the FTI ticket was resolved in a different way.

Please feel free to reopen.

**Pull request closed without merging**